core/arch/x86/decode_private.h

```diff
@@ -414,6 +430,7 @@ enum {
     TYPE_K_REG,                  /* modrm.reg selects k0-k7 */
     TYPE_K_VEX,                  /* vex.vvvv field selects k0-k7 */
     TYPE_K_EVEX,                 /* evex.aaa field selects k0-k7 */
+    TYPE_K_EVEX_V,               /* reg of modrm selects xmm/ymm/zmm with opmask */
```

Comment:
Type "K" is for opmask only. modrm reg field for xmm/ymm already exists, called type "V". Even though this is not officially spelled out like this in the Intel Specs., this is a natural extension of type "V".

Reply:
You are right the modrm reg field already exists. This field used for representing zmm register with opmask for example zmm1{K1}{z}.

Reply:
Each type is for a distinct interpretation of one bit-field. zmm1{K1}{z} implies 3 fields (one for zmm, one for k1, another bit for 'z') if needed. You'd add a type if there is no one available yet. In this case, everything is already there: For example modrm.reg with EVEX.RR' for zmm (see each individual instruction in Intel's Spec.). k zeroing or merge-mask should (almost-?) always be TYPE_K_EVEX (EVEX.aaa). There is probably no type needed for {z} (merge-mask bit).

Reply:
TYPE_K_EVEX_V  is merge of TYPE_K_EVEX and TYPE_V. For {z} field there is no meter to add type defination. With this way I can decide wheter operand of instruction has opmask or not?
Whether I understand to you correctly? You suggest to instead of defining a new type just use TYPE_V and add K* registers decoding evex.aaa.

Reply:
There need to be a new size _evex64 types like the existing ones for _vex32 that sould be used in combination with TYPE_V. TYPE_K is for mask registers only, there's no mixup of different register types in one type - yet. I think you're running into this because EVEX adds another mask operand. We might have to add another column to instr_info_t. Any other ideas?

Reply:
I think adding new column in instr_info_t is not good idea, because it will bring extra memory consumption (most of instructions has not opmask field). 
I  suggest two versions.
1) Because of destination operand of instructions can be zmm1{k1} (pair of modrm register and opmask register) then I suggest to define a new type which will represent combination of these registers, for example TYPE_K_EVEX_V. 
2) We may use types of VEX instructions and durring disassemble add k* register if evex.aaa was set (z field too). For  INSTR_CREATE_* we should add new iterface which will contain info about opmask. For example vmovupd instruction will have this defination
 #define INSTR_CREATE_vmovupd(dc, d, s) instr_create_1dst_k_1src((dc), OP_vmovupd, (d), (k), (s))

Reply:
Having merged types like TYPE_K_EVEX_V is not the way to go, the type is distinct for one bit-field. It would either be a new field in instr_info_t, or some kind of pointer to new table entry like pusha does it. Need to get quantified how many instructions there are with > 3 sources.

Reply:
Jivan, could you give us an idea by any chance who're you're working with / for, your long-term plans etc. Asking because as mentioned before, there will be / is considerable development for AVX-512 now at Google, these basic things here will overlap significantly. Once these basic things will be done, work can be more easily split up upstream, e.g. adding new instructions to the tables.

Reply:
There is an existing mechanism used to support instructions with more than 3 sources or more than 2 destinations: the extra_operands[] table holds a further entry (or 2 further entries if needed).  If we're adding a smallish number of such many-source instructions we should use that existing mechanism.  If we're talking about hundreds then maybe we would consider a new way to do it if we feel the extra entry is too awkward or inefficient: are there hundreds of AVX-512 instructions with 4 or more sources?

Reply:
A first scan through the Specs looks like that 3 sources should mostly be enough.

Reply:
@hgreving2304 I also think thant 3 sources and 2 destinations should mostly be enough.
For zmm1{k1} case may we use the first destination of instr_info_t for zmm1 and the second destination for k1 registers? If "yes" then I think we not need to define a new type. We can just use TYPE_V for the first destination (zmm*)  and TYPE_K_EVEX for the second destination (k*).

Reply:
@derekbruening Thank you for your suggestion. How I saw most part of insteuctions have less than 4 sources.

Reply:
@JivanH yes, except the mask register is not a destination, hence needs to be a source. Another friendly reminder to synch with us about your future near term plans and your background in more detail. Reason again is that here is R&D at Google that will otherwise likely merge patches coming soon that in part will resemble what you're doing (in addition to various other missing pieces in AVX-512). We're open to splitting up in particular the various instruction table extensions but this needs to be well coordinated. Also glad to have a phone conversation about it.

Reply:
Hendrik I work in ISP RAS. I going on to contribute with you for a month (for this patch). During that period of time I will add support for insteuctions with 0x0F implied leading opcode bytes. Of course the most part of time will take constructing architectire. Adding support/info of new insteuctions in tables is needs less effort. (This is my personaly opinion)